src/registry/mod.rs

```diff
@@ -14,44 +14,161 @@
 
 use std::{
     fmt::{self, Display},
-    ops::{Deref, DerefMut},
     str::FromStr,
 };
 
 mod artifactory;
 #[cfg(test)]
 mod cache;
 
-pub use artifactory::Artifactory;
+use crate::manifest::DependencyManifest;
+use crate::{config, manifest::Dependency};
+pub use artifactory::{Artifactory, CertValidationPolicy};
 use miette::{ensure, miette, Context, IntoDiagnostic};
 use semver::VersionReq;
 use serde::{Deserialize, Serialize};
 use thiserror::Error;
 use url::Url;
 
-use crate::manifest::{Dependency, DependencyManifest};
-
 /// A representation of a registry URI
 #[derive(Debug, Clone, Hash, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
 pub struct RegistryUri(Url);
 
+impl RegistryUri {
+    /// Get the host component of the registry URI
+    pub fn host(&self) -> Option<&str> {
+        self.0.host_str()
+    }
+
+    /// Get the path component of the registry URI
+    pub fn path(&self) -> &str {
+        self.0.path()
+    }
+}
+
+/// A reference to a registry
+#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub enum RegistryRef {
+    /// A URL to a registry
+    Url(RegistryUri),
+    /// An alias to a registry
+    Alias(String),
```

Comment:
In my understand you have two types:

RegistryRef (a reference, potentially aliased registry) and RegistryUri the uri version, independent of its origin

Reply:
You mean we could express the Alias as `Alias(String, Option<RegistryUri>)` ?

I found it safer to not use the option, but a third type. But for aliases, we need to remember the alias name and URL at all times, as, depending on the context, we do need to be able to access the alias, even if resolved.

What could be done is to split this enum into two types:

```rust
enum RawRegistryRef {
    Url(RegistryUri),
    Alias(String),
};

enum RegistryRef {
    Url(RegistryUri),
    Alias(String, RegistryUri),
};
```

Reply:
Well I'm talking about:

```rust
enum RegistryRef {
    Uri(Uri)
    Alias(RegistryName)
}

impl RegistryRef {
    fn resolve(config: &Config) -> Result<RegistryUri> {
        // 
    }
}

struct RegistryUri(Uri);
```

Reply:
Why do you need to remember aliases here?

Like you can do the resolving as late as possible (and still keep the original registry ref around)

Reply:
@mara-schulke : because resolution requires access to the config, and if postponing it, one has to pass the config reference into code that definitely shouldn't have to know about it.

Reply:
Let's push changing this to a follow-up PR, ok?